quiz/quiz.go

```diff
@@ -0,0 +1,182 @@
+a Quiz from file.
+*/
+package quiz
+
+import (
+	"bufio"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"text/tabwriter"
+	"time"
+)
+
+/*
+	This Quiz struct is a collection of up to 100 Questions, the name of the file containing the quiz, the number of
+questions in the quiz and the number of questions the user got correct.
+*/
+type Quiz struct {
+	Questions        [100]Question
+	Filename         string
+	TotalQuestions   int
+	CorrectQuestions int
+}
+
+/*
+	Creates a new Quiz struct based on the information in the file at the given file name
+*/
+func NewQuiz(filename string) (quiz Quiz, err error) {
+	fmt.Println("\nGetting quiz using file:\t", filename)
+
+	// open the file and save question to a Quiz struct
+	var questionCount = 0
+	csvFile, _ := os.Open(filename)
+	reader := csv.NewReader(bufio.NewReader(csvFile))
+
+	// loop until end of file or error is reached
+	for {
+		record, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			fmt.Println("Error reading from file:\t", err)
+			return Quiz{}, err
+		}
+
+		quiz.Questions[questionCount] = NewQuestion(record[0], record[1])
+		questionCount++
+	}
+
+	quiz.TotalQuestions = questionCount
+	quiz.CorrectQuestions = 0
+	quiz.Filename = filename
+
+	return quiz, nil
+}
+
+/*
+	Print prints a quiz to stdout with nicely aligned columns thanks to the tabwriter package. Useful for debugging.
+*/
+func (q Quiz) Print() (err error) {
+	fmt.Println("\nQuiz - ", q.Filename)
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 0, 8, 2, '\t', tabwriter.AlignRight)
+	_, err = fmt.Fprintln(w, "Question\tAnswer\tUser Answer\tCorrect")
+	if err != nil {
+		return err
+	}
+
+	for index, value := range q.Questions {
+		question := q.Questions[index]
+		if question.QuestionText != "" {
+			_, err = fmt.Fprintln(w, value)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	_, err = fmt.Fprintln(w)
+	if err != nil {
+		return err
+	}
+
+	err = w.Flush()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/*
+	Start begins the user-input portion of the quiz. Each question is presented to the user and they enter answers via
+stdin with "\n" being used to specify the end of an answer.
+*/
+func (q Quiz) Start(timer int) (Quiz, error) {
+	var err error
+
+	c := make(chan string)
+	successString, returnedQuiz := takeAndTimeQuiz(&q, timer, c)
+	if successString == "success" {
+		return *returnedQuiz, err
+	} else {
+		return *returnedQuiz, errors.New("quiz timed out before user completed it")
+	}
+
+}
+
+/*
+	takeAndTimeQuiz proctors the quiz, asking the user questions and saving the responses.
+	A Quiz pointer is used as input and output so we can be sure we're looking at the same quiz and not a copy.
+    timer is the time in seconds to allow the quiz to run for. If the user doesn't finish in time, the rest of the answers
+    are considered incorrect.
+*/
+func takeAndTimeQuiz(q *Quiz, timer int, c chan string) (string, *Quiz) {
```

Comment:
I'd recommend changing the name 'timer' here to something akin to 'maxQuizTime'

Reply:
Maybe you could call it timeout and have it be a `time.Duration`.

Additionally I would recommend making this an unexported method which wold allow you to remove the quiz parameter and return value.

Also on parameters, I'm not seeing the reason for passing in a channel here.